### PR TITLE
Adding ExactIntegrationUtility to the core

### PR DESCRIPTION
**Pull request by loumalouomega**
@andreas-apostolatos and @philbucher ask me about my exact integration utility. I spoke with @RiccardoRossi and we decided to move it to the core in order everyone can use the utility. Additionally I added some mortar utilities and classes

### CODE REVIEW

**Comment by loumalouomega** [comment]
I added mapping variables, that way we can declare common variables between applications

**Comment by loumalouomega** [comment]
The utility has been cleaned and optimized during this weekend. Some bugs were found last week and now the Utility is able to correctly integrate double curvature surfaces. There is a test in my branch for that, unfortunately I will not commit to the core, because it depends of other utilities (interface and search basically)

**Comment by loumalouomega** [comment]
I will shoe here a small example, but once it is merged with the core I will add it to the wiki

**Comment by loumalouomega** [comment]
For example what I do in https://github.com/KratosMultiphysics/Kratos/blob/features_mortar_contact_advances/applications/ContactStructuralMechanicsApplication/custom_conditions/mesh_tying_mortar_condition.cpp
~~~cpp
ExactMortarIntegrationUtility<TDim, NumNodes> integration_utility = ExactMortarIntegrationUtility<TDim, NumNodes>(mIntegrationOrder);
    
    // Create and initialize condition variables:
    GeneralVariables rVariables;
    
    for (auto it_pair = all_conditions_maps->begin(); it_pair != all_conditions_maps->end(); ++it_pair )
    {
        Condition::Pointer pCond = (it_pair->first);
        
        // Reading integration points
        ConditionArrayListType conditions_points_slave;
        const bool is_inside = integration_utility.GetExactIntegration(this->GetGeometry(), this->GetValue(NORMAL), pCond->GetGeometry(), pCond->GetValue(NORMAL), conditions_points_slave);
        
        if (is_inside == true)
        {
            GeometryType::IntegrationPointsArrayType all_integration_points_slave;
            
            for (unsigned int i_geom = 0; i_geom < conditions_points_slave.size(); i_geom++)
            {
                std::vector<PointType::Pointer> points_array (TDim); // The points are stored as local coordinates, we calculate the global coordinates of this points
                for (unsigned int i_node = 0; i_node < TDim; i_node++)
                {
                    PointType global_point;
                    GetGeometry().GlobalCoordinates(global_point, conditions_points_slave[i_geom][i_node]);
                    points_array[i_node] = boost::make_shared<PointType>(global_point);
                }
                
                DecompositionType decomp_geom( points_array );
                
                const bool bad_shape = (TDim == 2) ? ContactUtilities::LengthCheck(decomp_geom, this->GetGeometry().Length() * 1.0e-6) : ContactUtilities::HeronCheck(decomp_geom);
                
                if (bad_shape == false)
                {
                    const GeometryType::IntegrationPointsArrayType& integration_points_slave = decomp_geom.IntegrationPoints( this_integration_method );
                    
                    // Integrating the mortar operators
                    for ( unsigned int point_number = 0; point_number < integration_points_slave.size(); point_number++ )
                    {
                        const double weight = integration_points_slave[point_number].Weight();
                        const PointType local_point_decomp = integration_points_slave[point_number].Coordinates();
                        PointType local_point_parent;
                        PointType gp_global;
                        decomp_geom.GlobalCoordinates(gp_global, local_point_decomp);
                        GetGeometry().PointLocalCoordinates(local_point_parent, gp_global);
                        
                        const double DetJ = decomp_geom.DeterminantOfJacobian( local_point_decomp );
                        
                        all_integration_points_slave.push_back( IntegrationPointType( local_point_parent.Coordinate(1), local_point_parent.Coordinate(2), weight * DetJ ));
                    }
                }
            }
 
            if (all_integration_points_slave.size() > 0)
            {
                mPairSize += 1;
                mThisMasterConditions.push_back(pCond);
                mIntegrationPointsVector.push_back(all_integration_points_slave);
            }
        }
    }
~~~

**Comment by loumalouomega** [comment]
This is the test of double curvature I am mentioning (https://github.com/KratosMultiphysics/Kratos/blob/features_mortar_contact_advances/applications/ContactStructuralMechanicsApplication/tests/test_double_curvature_integration.py):

imagen

Plotting the result with Mathematica it results in:

imagen

Where the red surfaces are the Slave conditions and the blue are the Master conditions. The semi-transparent triangles correspond with the integration triangles. It includes the IDs of the conditions. You can check in more detail in: 
[double_curvature_test.pdf](https://github.com/KratosMultiphysics/Kratos/files/1170635/double_curvature_test.pdf)

To get the Mathematica code, you can activate the debug, look at the line 89 from https://github.com/KratosMultiphysics/Kratos/blob/features_mortar_contact_advances/applications/ContactStructuralMechanicsApplication/tests/test_double_curvature_integration.py

**Comment by philbucher** [comment]
Thanks a lot Vicente!

File: kratos/python/add_mapping_variables_to_python.cpp
Comment (RiccardoRossi) [inline kratos/python/add_mapping_variables_to_python.cpp]: are u sure that this macro magic is needed?

@roigcarlo opinions on this?

**Review by RiccardoRossi — APPROVED** [review]
looks fine to me.

the only question is that you see that you have some "del" in the python script. Why do you need to do that explicitly?

**Comment by pooyan-dadvand** [comment]
@loumalouomega please don't merge it yet...

File: kratos/python/add_mapping_variables_to_python.cpp
Comment (roigcarlo) [reply kratos/python/add_mapping_variables_to_python.cpp]: Dosn't seem to do anything to be honest. Also, I spoke with @pooyan-dadvand and it looks that this is some kind of bad replace ( name is veeeery suspicious ).

The macro also appears in:
- add_ale_variables_to_python.cpp
- add_cfd_variables_to_python.cpp
- add_legacy_structural_variables_to_python.cpp

And also does nothing. I would say is safe to remove, and also create an issue to remove it from the other files as well.

**Comment by loumalouomega** [comment]
Sorry for the macros, I just copied the ALE variables file, that I assume copied the CFD. We can use this PR to remove unnecessary macros

File: kratos/includes/mapping_variables.h
Comment (philbucher) [inline kratos/includes/mapping_variables.h]: Isn't that sth from the contact application?
Or do you mean neighboring pairs?

Comment (loumalouomega) [reply kratos/includes/mapping_variables.h]: Are the neighbouring, we can change the name for being something more general

Comment (philbucher) [reply kratos/includes/mapping_variables.h]: I think this would be more clear
I usually use names with "neighbor" in them

**Comment by loumalouomega** [comment]
I applied suggested changes

File: kratos/includes/mapping_variables.h
Comment (jcotela) [inline kratos/includes/mapping_variables.h]: A minor comment, but please refrain from using accented characters (or non-ascii in general) in source code files, since it tends to produce errors when changing between windows and linux.

Comment (loumalouomega) [reply kratos/includes/mapping_variables.h]: Sorry, I will correct it

Comment (loumalouomega) [reply kratos/includes/mapping_variables.h]: Done

**Comment by pooyan-dadvand** [comment]
Is there any way to reduce number of arguments in this method?

```c++
 const bool is_inside = integration_utility.GetExactIntegration(this->GetGeometry(), this->GetValue(NORMAL), pCond->GetGeometry(), pCond->GetValue(NORMAL), conditions_points_slave);
```

**Comment by loumalouomega** [comment]
The normal of the master can be avoided in the 3D, but it is a template...

**Comment by loumalouomega** [comment]
Well, we can use the Normal() that is now implemented in the geometry, but that means to recompute again. We can do that if you agree @pooyan-dadvand

File: kratos/utilities/exact_mortar_segmentation_utility.h
Comment (philbucher) [inline kratos/utilities/exact_mortar_segmentation_utility.h]: Typo?

Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h]: Yep, I have pending anyway to add my last modifications to this PR

Comment (loumalouomega) [reply kratos/utilities/exact_mortar_segmentation_utility.h]: In fact I corrected this in my last modifications

**Comment by loumalouomega** [comment]
Added the last update, this modification is to detect where the intersection belongs (I use it to define the derivatives of the segmentation)

**Comment by pooyan-dadvand** [comment]
I would recalculating the normal also to avoid mistakes in passing a wrong normal. @RiccardoRossi what is your opinion about its external cost?

**Comment by loumalouomega** [comment]
@RiccardoRossi Ping

**Comment by loumalouomega** [comment]
@RiccardoRossi ping

**Comment by loumalouomega** [comment]
Now that the ContactStructuralMechanicsApplication is merged to the master I will clean the application in this branch to avoid duplication of the app

**Comment by loumalouomega** [comment]
Donem @RiccardoRossi ping

**Pull request merged**
